### ISSUE TITLE
Cluster-mode: request blocked by busy worker even though other worker has idle threads

### ISSUE THREAD

**Issue by sbull** [puma/puma#1107]
I'm exploring cluster mode (with MRI, linux) with Rails 5, and I'm seeing some unfortunate behavior. I'm using 2 worker processes with 5 threads each (like the defaults in puma-heroku), with `preload_app!`. I've intentionally made a "slow" controller action to play with the thread & process operation and check how the requests are handled under load.

The mechanism by which puma chooses which worker in a cluster will handle a request is unclear to me (I'm thinking it might be up to the OS for whichever process grabs the IO first in Puma::Server#handle_servers?), but in my dev environment it seems that one of the workers is usually favored over the other - significantly different from a random distribution (I see the worker PID in my logs). So, I make 5 concurrent requests for the "slow" action, and on the 6th concurrent request to a "fast" action, it blocks until one of the "slow" requests finishes, and is handled by the same worker. But on my 7th concurrent request to a "fast" action, still before any of the first 6 requests have finished, it gets handled by the other worker without delay.

My conclusion is that worker#0 must be taking requests and handing them to its available threads until it determines that it has no more threads ready to handle them, but it doesn't figure that out until AFTER it's already grabbed the 6th request, preventing worker#1 from servicing it. So even though I've got a whole extra worker of capacity with 5 idle threads, my poor request 6 gets blocked until one of worker#0's threads frees up.

Is there a way to get around this? Maybe worker#0 could stop trying for the IO when all of its threads are busy - or maybe this is something related to the `queue_requests` option?

Here's the templated info:
### Steps to reproduce

1) Make a "slow" action along with a "fast" action:

``` ruby
class TesterController < ApplicationController

  def slow
    sleep 20
    render plain: 'slow'
  end

  def fast
    render plain: 'fast'
  end

end
```

2) Make 5 concurrent requests to /slow, then a concurrent request to /fast, and another concurrent request to /fast. Check which worker handles the requests (in log files, or debug messages?), and note the timing.
### Expected behavior

When the first 5 concurrent requests to /slow are handled by worker#0, both requests to /fast should be handled by worker#1.
### Actual behavior

When the first 5 concurrent requests to /slow are handled by worker#0, the 6th request to /fast is handled by worker#0 as well, and waits until a /slow request finishes before being processed. The 7th request to /fast appears to be correctly handled by worker#1.
### System configuration

**Ruby version**: MRI 2.3.1 (via rbenv)
**Rails version**: 5.0.0.1

**Comment by ce07c3** [comment]
I might be having similar issues. Under load (and there are a few slow controller actions), requests get blocked.

**Comment by amichal** [comment]
we had this too in an older version, `queue_requests` appears to fix it

**Comment by chewi** [comment]
Me too. But @amichal, `queue_requests` is enabled by default and disabling it worked for me. Is that what you meant?

**Comment by mitchenkod** [comment]
I have faced this issue too. The file uploading process, when a few relatively large files (~ 150 mb) are uploading concurrently puma server becomes completely unresponsive. This is the issue, I have closed it because of duplicating https://github.com/puma/puma/issues/1115.

**Comment by nateberkopec** [comment]
> The mechanism by which puma chooses which worker in a cluster will handle a request is unclear to me

I'm pressed for time right now and can't dig into this issue in full, but Puma uses the `select` system call to manage this, so `$ man select` may give some insight here.

**Comment by nateberkopec** [comment]
Also, it's very relevant that your test controller using `sleep`. Sleep unblocks the MRI interpreter and allows other threads to do work. You will probably get different results if your controller looks like the following:

```ruby
def slow
    time = Time.now
    while Time.now <= time + 20
      # do nothing, spin CPU cycles
    end
    render plain: 'slow'
  end
```

**Comment by chewi** [comment]
I mostly saw this while doing large database imports, which I think would exhibit the same behaviour as `sleep `.

**Comment by nateberkopec** [comment]
Closed by #1278
